schemachanger: implement `ADD CONSTRAINT ... FOREIGN KEY`

Xiang-Gu opened the pull request:
This PR implements `ADD CONSTRAINT ... FOREIGN KEY` in the declarative schema changer.

One notable change is the behavioral difference compared to legacy schema changer: we only
add the FK constraint onto the mutation slice, and when it's finally ready to get published after
validation succeeded, we remove it off the mutation slice, add it to `OutboundFKs` slice, 
and add the back-reference in the referenced table's `InboundFKs` slice.

See each commit message for more details.

Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93068)
<!-- Reviewable:end -->

Xiang-Gu committed:
opgen: adding support for the adding/dropping path of ForeignKey element

Xiang-Gu committed:
backfill: Add logic for foreign key validation

This commit refactored some of the check constraint validation code
to implement a unified function for validating all kinds of constraints.

Xiang-Gu committed:
scbuildstmt: Implement `ADD FOREIGN KEY` in the builder

Xiang-Gu committed:
schemachanger: Enable `ADD FOREIGN KEY` by default and add tests

Xiang-Gu committed:
rewrite: Rewrite FK in mutations during restore

Xiang-Gu commented:
This is ready for a look! Overall, It follows the same recipe as `ADD CHECK`. One interesting thing is backup/restore test revealed a TODO in the `rewrite.go` and commit 5 fixed it.

postamar commented on pkg/sql/catalog/table_elements.go:
```diff
@@ -440,6 +440,9 @@ type Constraint interface {
 	GetName() string
 
+	// GetConstraintType returns the constraint type.
+	GetConstraintType() descpb.ConstraintType
```
Rather than make this a method on the `Constraint` interface, can you make this a function instead? This isn't terribly important in this case, but a useful general principle is to keep the set of object methods to a minimum and to keep separate any logic which can be expressed using method calls and doesn't need any state that's internal to the object.

postamar commented on pkg/sql/backfill.go:
```diff
@@ -1551,10 +1552,38 @@ func ValidateCheckConstraint(
+			)
+		} else {
+			return errors.AssertionFailedf("validation of unsupported constraint type")
+		}
```
very minor nit: is it possible to `switch constraint.(type)` here instead? I'm not actually sure, in any case these if-elses feel kinda clunky.

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/constraint.go:
```diff
@@ -169,20 +169,125 @@ func (m *visitor) RemoveForeignKeyConstraint(
+		return errors.AssertionFailedf("foreign key with ID %d not found in origin table %q (%d)",
+			op.ConstraintID, out.GetName(), out.GetID())
+	}
+	return nil
```
nit: fyi you could probably do away with `found` and `break` etc. with a `defer`, same applies to the ops below; not sure if it's worth it though. This is fine btw.

postamar commented on pkg/sql/schemachanger/scpb/BUILD.bazel:
```diff
@@ -17,6 +17,7 @@ go_library(
     deps = [
         "//pkg/sql/catalog/catpb",  # keep
         "//pkg/sql/sem/catid",  # keep
+        "//pkg/sql/sem/tree",  # keep
```
No, please don't import `tree`! In fact, the enums you need are already in `catpb`, use those instead.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -141,6 +150,249 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+	if originalTableNamespaceElem.DatabaseID != referencedTableNamespaceElem.DatabaseID {
+		// TODO (xiang): Refactor new schema changer to be able to retrieve cluster settings (sql.allowCrossDatabaseFKs)
+		// here in the builder.
+		panic(scerrors.NotImplementedErrorf(t, "foreign key references cross databases are not supported yet"))
```
Cross-DB references are a deprecated feature, effectively already unsupported, so you can do away with this and simply throw a regular error.

postamar reviewed (changes requested):
Very nice!

I did a preliminary review and couldn't find too much to complain about.

Xiang-Gu committed:
catenumpb: Moved some enums into a new low-dependency package

Previously, we had enum protobuf in catpb for FK reference actions but
written in proto2, making it impossible to re-use in other proto3 files
(e.g. scpb). This commit solved this problem by moving the following two
enum definitions into a new package `sql/sem/semenumpb/constraint.proto`,
written in proto3:
  - catpb.proto: ForeignKeyAction
  - descpb.proto: ForeignKeyConstraint_Match

The idea of this package is that it will be imported by tree and possibly
other XXpb packages, so that all the other XXpb packages do not need to
import the heavy tree package.

It also did some cleanup work to move some other enum
definitions out of `catpb` into a new package
`sql/catalog/catpb/enum.proto`, also in proto3, for future use:
  - catpb.proto: SystemColumnKind
  - catpb.proto: InvertedIndexColumnKind

The general rule of thumb to determine which proto file an enum should
go to is this: If it's a concept that is visible to the SQL layer (e.g.
ON UPDATE, MATCH, etc. that has keyword and can be configured in a SQL
query) should go to `sql/sem/semenumpb/xxx.proto`. Other concepts that
are internal (e.g. SystemColumnKind) should be put in
`sql/catalog/catpb/enum.proto`.

Xiang-Gu reviewed (commented):
@postamar Thanks for your feedback. I've addressed your feedback and pushed the changes. One major change is that I re-defined the concepts of ForeignKeyReferenceAction and ForiegnKeyReferenceMatchMethod inside `elements.proto` due to the reasons discussed in this [thread](https://cockroachlabs.slack.com/archives/CUQDCNZ46/p1671563378370929). 

I'd appreciate it if you can take another look at it when you get a chance.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-:-NJlgHnJ8nll15O9Gude:b-osjnr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/backfill.go` line 1586 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXVO63cY9NqL97L-r2-1586:-NJkRyXo-To0LxEnACqX:b-4ta7vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd4080b311f8302cd1f080376e94dcaa04be5810/pkg/sql/backfill.go#L1586)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

very minor nit: is it possible to `switch constraint.(type)` here instead? I'm not actually sure, in any case these if-elses feel kinda clunky.
</blockquote></details>

I don't think that's possible bc constraint.(type) will be `*tabledesc.checkConstraint`, `*tabledesc.foreignKeyConstraint`, etc., all of which are unexported.

However I do agree with you that this if-elses feel kinda clunky so I will use the newly added function GetConstraintType and switch on various `descpb.ConstraintType`.
___
*[`pkg/sql/catalog/table_elements.go` line 444 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NJM7gwteq9O5pzXxSyM-r1-444:-NJkjGGM23rUnz4cugOq:b-1qxhxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/880d380409847033de63289a7657a37494a370df/pkg/sql/catalog/table_elements.go#L444)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

FYI, I looked into where this new method is used and it doesn't seem like it's particularly necessary. There's a bunch of type checks such as `c.ConstraintType == descpb.ConstraintToUpdate_FOREIGN_KEY` which can (or rather, should) be rewritten as `c.AsForeignKey() != nil` and there's a format string parameter which could be replaced by a `%T` or something. 
</blockquote></details>

Done. I removed this method from the interface and instead created a function that takes in a `c Constraint` and returns its `descpb.ConstraintType`.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r4-234:-NJkkWkg1uPb747Jv0k-:b-7oieq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa7f7e1ea9f2e0116f6a90613cd32ed2f5f3d858/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L234)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Cross-DB references are a deprecated feature, effectively already unsupported, so you can do away with this and simply throw a regular error.
</blockquote></details>

good to know! Done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 588 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r4-588:-NJklvsNBUaame8OQ73t:bnp2pls) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa7f7e1ea9f2e0116f6a90613cd32ed2f5f3d858/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L588)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Don't we have filters which filter element result sets by attribute value? Seems like you could use an existing one or define a new one.
</blockquote></details>

That's correct! I wasn't aware of them and I have made the changes.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/constraint.go` line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIiw2rtPkaLLPwtsVuk-r3-193:-NJkngkgEJVPP7OOobLk:bjqkn5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/af6528d54324a7e0f7ab744ca63398f343032b04/pkg/sql/schemachanger/scexec/scmutationexec/constraint.go#L193)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: fyi you could probably do away with `found` and `break` etc. with a `defer`, same applies to the ops below; not sure if it's worth it though. This is fine btw.
</blockquote></details>

Can you tell me a bit more about how to replace this pattern with a `defer`? 
___
*[`pkg/sql/schemachanger/scpb/BUILD.bazel` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIiw2rusngwrZ_zo8Y1-r3-20:-NJkor7zDj3QWVH7wAfC:b-n0ub9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/af6528d54324a7e0f7ab744ca63398f343032b04/pkg/sql/schemachanger/scpb/BUILD.bazel#L20)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

No, please don't import `tree`! In fact, the enums you need are already in `catpb`, use those instead.
</blockquote></details>

This is a choice of an unfortunate situation: I also wanted to use the enums in `catpb` when I introduced those fields in `scpb.ForeignKeyConstraint`. But I got an error because `catpb` uses proto2 and `scpb` is proto3, and we can not use enum values defined in proto2 directly in proto3 (https://developers.google.com/protocol-buffers/docs/proto3#using_proto2_message_types)

Please let me know if you know an alternative solution here. I also don't like importing `tree` here but can not think of another way to do this.



_Code quote:_
```Starlark
        "//pkg/sql/catalog/catpb",  # keep
        "//pkg/sql/sem/catid",  # keep
        "//pkg/sql/sem/tree",  # keep
```


<!-- Sent from Reviewable.io -->

ajwerner commented:
I agree with not importing tree into a proto package. Some options:

* use a wrapper message defined in proto2 and embed it
* lift the enum to go completely and keep it defined in the same package. Use a cast type

mgartner commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -2715,9 +2715,16 @@ statement ok
 ALTER TABLE child_duplicate_cols ADD CONSTRAINT fk FOREIGN KEY (parent_id, parent_id) references parent
 
+skipif config local-legacy-schema-changer
+statement error pq: foreign key contains duplicate column "test.public.child_duplicate_cols.parent_id"
```
Can you test for the correct pgcode here? Postgres uses `42830`.

Xiang-Gu replied:
done

mgartner commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -2715,9 +2715,16 @@ statement ok
 
 # The fk constraint is invalid because it has duplicate columns, so automatically adding the index fails
+# Unfortunately, the error message is slightly different between legacy and declarative schema changer
+# so we have the same statement twice here.
```
Should we leave a TODO to fix this? Seems like the fully qualified name is a bit over the top in this case.

postamar replied:
Use a regexp that matches both schema changer implementations.

mgartner commented on pkg/sql/sem/tree/constants.go:
```diff
@@ -10,6 +10,90 @@
+
+// ReferenceActions contains the actions specified to maintain referential
+// integrity through foreign keys for different operations.
+type ReferenceActions struct {
```
Why did you move this type and related global variables to this file?

Xiang-Gu commented:
@postamar Sorry for the delayed response. I was trying to make the first commit work where I did a little bit of refactoring work. What I essentially did is that I pulled the FkReferenceActio and FKCompositeMatchMethod from catpb and descpb into a new proto file (called `catenumpb` with proto3 syntax). Its benefits are two-folds:
1. We can now use those enum types in scpb (proto3) so we don't need to import `tree` in scpb.
2. It makes the dependency cleaner to benefit future changes. Surprisingly, we already have [one such PR](https://github.com/cockroachdb/cockroach/pull/94116) that would potentially benefit from this separation.

Yes, while the CI is running and I'm actively on top of it in case it fails, please give me another round of review when you get a chance! Thank you very much!

postamar commented:
Looks like the build is broken, `dev test pkg/sql/catalog/tabledesc` should reveal problems.

Xiang-Gu commented:
@postamar I'd glad that I finally made the CI green again. I spent some time trying to separate some of the enum msg definitions (written in proto2) into a new file (written in proto3). It will be a low-level package so package `scop` does not need to depend on package `tree`, and we can workaround the enum-in-proto2-cannot-be-used-in-proto3 limitation. I think this is ready for another round of review. Thank you in advance!

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -321,6 +321,8 @@ message UniqueWithoutIndexConstraint {
   uint32 constraint_id = 2 [(gogoproto.customname) = "ConstraintID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.ConstraintID"];
   repeated uint32 column_ids = 3 [(gogoproto.customname) = "ColumnIDs", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.ColumnID"];
+  // Predicate, if non-nil, means a partial uniqueness constraint.
+  Expression predicate = 4 [(gogoproto.customname) = "Predicate"];
```
This doesn't seem like it should be in commit b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -141,6 +150,248 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+	// 5. Resolve `t.(*ForeignKeyConstraintTableDef).Table` (i.e. referenced table name) and check whether it's in the same
+	//    database as the originTable (i.e. tbl). Cross database FK references is a deprecated feature and is in practice
+	//    no longer supported. We will return an error here directly.
+	referencedTableID := mustGetTableIDFromTableName(b, fkDef.Table)
```
`mustGetTableIDFromTableName` requires `privilege.CREATE` to resolve the table, which seems odd for a table referenced in a FK. Can you double-check that this is the existing behaviour, please?

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -141,6 +150,248 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+
+	// 5. Resolve `t.(*ForeignKeyConstraintTableDef).Table` (i.e. referenced table name) and check whether it's in the same
+	//    database as the originTable (i.e. tbl). Cross database FK references is a deprecated feature and is in practice
+	//    no longer supported. We will return an error here directly.
```
style nit: Comments shouldn't go beyond the 80th column. Applies to other comments in this file also.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -141,6 +150,248 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+	originalTableNamespaceElem := mustRetrieveNamespaceElem(b, tbl.TableID)
+	referencedTableNamespaceElem := mustRetrieveNamespaceElem(b, referencedTableID)
+	if originalTableNamespaceElem.DatabaseID != referencedTableNamespaceElem.DatabaseID {
+		panic(scerrors.NotImplementedErrorf(t, "cross DB FK reference is a deprecating feature and won't be supported"+
```
s/deprecating/deprecated/
s/won't be supported in declarative schema changer./is no longer supported/

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -141,6 +150,248 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+
+	// 7. Compute referencedColumnNames, which is usually provided in `t.(*ForeignKeyConstraintTableDef).ToCols`. But if
+	//    it's empty, then we attempt to add this FK on the PK of the referenced table, excluding implicit columns.
+	if fkDef.ToCols == nil || len(fkDef.ToCols) == 0 {
```
just `len(fkDef.ToCols) == 0`

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -141,6 +150,248 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+		}
+	}
+
+	// 10. Check the name of this to-be-added FK constraint hasn't been used; Or, give it one if no name is specified.
```
s/Check/Check that/

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go:
```diff
@@ -141,6 +150,248 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+		))
+	}
+
+	// 12. (Finally!) Add a ForeignKey_Constraint, ConstraintName element to builder state.
```
Thanks these comments are very helpful!

postamar reviewed (approved):
LGTM, just minor issues which should be straightforward to resolve. Great work!

Xiang-Gu reviewed (commented):
Thanks for the review! I've addressed the feedback and will merge it once CI is green.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-:-NLbkaDX5dMP3lnf1hNg:b-joxe3h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @postamar)

___
*[`pkg/sql/backfill.go` line 1586 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXVO63cY9NqL97L-r2-1586:-NLau0554qfzADyxJU1w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd4080b311f8302cd1f080376e94dcaa04be5810/pkg/sql/backfill.go#L1586)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

> I don't think that's possible bc constraint.(type) will be *tabledesc.checkConstraint, *tabledesc.foreignKeyConstraint, etc., all of which are unexported.

FYI you can `case catalog.CheckConstraint:` etc but this is fine too of course.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 227 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r34-227:-NLazf4gEFN3LPP6lVvX:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L227)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

style nit: Comments shouldn't go beyond the 80th column. Applies to other comments in this file also.
</blockquote></details>

Done



_Code quote:_
```Go
 originTable (i.e. tbl). Cross database FK refer
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 228 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r34-228:-NLbexANBMrF8SHXyj2C:b-ahy0y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L228)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`mustGetTableIDFromTableName` requires `privilege.CREATE` to resolve the table, which seems odd for a table referenced in a FK. Can you double-check that this is the existing behaviour, please?
</blockquote></details>

I don't think there is an existing behaviour per se; It's just that we need to know a table ID from its name here, and the only utility we have in the builder is `ResolveTable` which requires a privilege. I put `privilege.CREATE` here bc `ADD FOREIGN KEY` requires `CREATE` privilege.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 232 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r34-232:-NLbg00FDo1qyRmk28PG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L232)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

s/deprecating/deprecated/
s/won't be supported in declarative schema changer./is no longer supported/
</blockquote></details>

Done.



_Code quote:_
```Go
ference is a dep
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 255 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r34-255:-NLbg6hc1lOMuU5djD-5:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L255)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

just `len(fkDef.ToCols) == 0`
</blockquote></details>

done



_Code quote:_
```Go
f.ToCols == nil || len(fkDef.ToCo
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 310 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r34-310:-NLbgDTM7wyJlriVbSGt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L310)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

s/Check/Check that/
</blockquote></details>

Done.



_Code quote:_
```Go
eck the name of this
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go` line 336 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkY8-r34-336:-NLbgEQD1TzxO61hfZbn:b-ajetbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go#L336)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Thanks these comments are very helpful!
</blockquote></details>

Good to hear!
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 325 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NIYnkXW1U7NH965SkYD-r34-325:-NLbi8M-8rgwyeI6htYl:b-vchrnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d/pkg/sql/schemachanger/scpb/elements.proto#L325)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This doesn't seem like it should be in commit b4eabafbde17fe4ffaf4fc74ea7d58fd346b508d
</blockquote></details>

This I believe it does! When implementing `ADD FOREIGN KEY` in the builder, we need to ensure that the referenced table guarantees uniqueness on the referenced columns. One way to guarantee uniqueness is via a *non-partial* unique_without_index constraint on the referenced columns. 
___
*[`pkg/sql/schemachanger/screl/attr.go` line 181 at r36](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NJRK2IZyYrmUskv7ElE-r36-181:-NLbgOUtCYZj3F2NAQC-:b-5zt63k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f490c82340ccd0769bed4fec481fc9a0ecf3de4/pkg/sql/schemachanger/screl/attr.go#L181)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I just noticed something: is this actually needed, can you check? if not, please remove this binding.
</blockquote></details>

Yes, I think so. Just like `CheckConstraint`, it's possible that a FK constraint validation needs to perform on a new primary index that's yet not public. Having this rel attribute allows the dep-rule "don't validate until the index is backfilled" to apply to `FK` constraints.
___
*[`pkg/sql/logictest/testdata/logic_test/fk` line 2719 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/93068#-NJRK2IZyYrmUskv7El3-r18-2719:-NLbgqv42DkcDVEJk8c0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d226ea35677eed260bf4d5025f3a228e23836f65/pkg/sql/logictest/testdata/logic_test/fk#L2719)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Use a regexp that matches both schema changer implementations.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
